### PR TITLE
Add Unit Tests for Audio Module

### PR DESCRIPTION
**Pull request by yl54**
## Description
- Added a few unit tests to test the audio library.
- Added version `0.5.0` of amethyst_utils to audio directory.

## Checklist
By placing an x in the boxes I certify that I have:

- [x] Ran `cargo test --all` locally if this modified any rs files.
- [x] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [x] Updated the content of the book if this PR would make the book outdated.
- [x] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [x] Added unit tests for new APIs if any were added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

**Comment by fhaynes** [comment]
Thank you! I will have to update the Dockerfile to include a /dev/null equivalent for audio, and then I'll re-run this.

File: amethyst_audio/src/sink.rs [hunk 1]
```diff
@@ -59,3 +59,62 @@ impl AudioSink {
         self.sink.stop();
     }
 }
+
+// q: for AudioSink, do we want to test each of these apis as a sanity check?
```
Comment (azriel91) [inline amethyst_audio/src/sink.rs]: I don't think it's necessary; it may be a better idea to do one of the following in a separate PR:

* implement `Deref` and `DerefMut` with `Sink` as the target type, or
* use the [`delegate`](https://crates.io/crates/delegate) crate -- not really liking this option, but worth listing

File: amethyst_audio/src/components/audio_emitter.rs [hunk 2]
```diff
@@ -53,3 +54,79 @@ impl AudioEmitter {
...
+        f.read_to_end(&mut buffer).unwrap();
+
+        // Create a Source and AudioEmitter from those bytes
+        let src = Source { bytes: buffer };
+        let mut emitter = AudioEmitter::default();
+
+        // Call play
+        match emitter.play(&src) {
+            Ok(_pass) => assert!(should_pass),
+            Err(_fail) => assert!(!should_pass),
```
Comment (azriel91) [inline amethyst_audio/src/components/audio_emitter.rs]: for the it-failed-but-should-not-fail case, since the error type contains information, perhaps do something like this:

```rust
e @ Err(_fail) => assert!(!should_pass, "Expected `play` result to be `Ok(..)`, but was {:?}", e),
```

(syntax might be wrong, but there's the idea)

File: amethyst_audio/src/components/audio_emitter.rs [hunk 3]
```diff
@@ -53,3 +54,79 @@ impl AudioEmitter {
 impl Component for AudioEmitter {
     type Storage = BTreeStorage<Self>;
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{AudioEmitter, Source};
+    use amethyst_utils::app_root_dir::application_root_dir;
+    use std::{fs::File, io::Read, vec::Vec};
```
Comment (azriel91) [inline amethyst_audio/src/components/audio_emitter.rs]: by the way, could you add empty newlines between import groups like this:

```rust
use std::{..};
// newline
use amethyst_utils::{..};
use another::{..};
// newline
use crate::{..};
```

we adopted the convention to quickly see what's from std / other crates / current crate.

oh ya, are you used to `use`ing `std::{boxed::Box, vec::Vec}`? (guessing that's standard practice for a `#[no_std]` crate?)
normally we leave them out as well

File: amethyst_audio/src/components/audio_emitter.rs [hunk 4]
```diff
@@ -53,3 +54,79 @@ impl AudioEmitter {
...
+    #[test]
+    fn test_picker() {
+        // Create the input variables
+        let mut emitter_main = AudioEmitter::default();
+        let box_picker: Box<dyn FnMut(&mut AudioEmitter) -> bool + Send + Sync> =
+            Box::new(use_audio_emitter);
+
+        // Test set_picker and assert that it is not empty
+        emitter_main.set_picker(box_picker);
+        assert!(!emitter_main.picker.is_none());
```
[the remark this answers is not in the document]
Comment (yl54) [reply amethyst_audio/src/components/audio_emitter.rs]: That sounds better I kinda like that instead of the weird not none.

File: amethyst_audio/src/components/audio_emitter.rs (hunk 3 above)
Comment (yl54) [reply amethyst_audio/src/components/audio_emitter.rs]: I saw the empty new lines in the other src files. Can change for tests as well.

I'm not used to anything I'm new to the language. Happy to help keep conventions consistent.

**Comment by fhaynes** [comment]
OK, this passes on linux now, but not Windows. I'll see if I can add a loopback audio device to Windows later today.

**Comment by fhaynes** [comment]
Hmm. I think we're going to have to only run these tests on linux for the moment. Windows is going to be a bit more challenging...

File: amethyst_audio/src/sink.rs (hunk 1 above)
Comment (yl54) [reply amethyst_audio/src/sink.rs]: Oops that was not meant for review. I was gonna ask in Discord or something. Will remove the comment.

File: amethyst_audio/src/components/audio_emitter.rs (hunk 2 above)
Comment (yl54) [reply amethyst_audio/src/components/audio_emitter.rs]: I added more useful commentary.

**Comment by yl54** [comment]
At least to mitigate the issue of adding a larger file, I can add a smaller wav file.

File: amethyst_audio/src/components/audio_emitter.rs (hunk 2 above)
Comment (azriel91) [reply amethyst_audio/src/components/audio_emitter.rs]: ah, could you either use `e`, or remove the underscore from `_fail` (underscore means it's unused).

Also, could you use the mp3 and wav from https://github.com/mathiasbynens/small ? (gotta rebase the bigger files out of history)
Thanks!

**Comment by azriel91** [comment]
ah, just saw your last push. Unfortunately replacing a file in git doesn't save us the space unless we squash commits -- which [our bot can't seem to d](https://bors.tech/documentation/getting-started/), so could you please rebase / squash the commits? thanks!

File: amethyst_audio/src/components/audio_emitter.rs (hunk 2 above)
Comment (yl54) [reply amethyst_audio/src/components/audio_emitter.rs]: Removed all instances of the fail underscores. ty for catching it.

**Comment by yl54** [comment]
Squashed everything.

**Comment by azriel91** [comment]
bors r=torkleyy, fhaynes, azriel91

**Comment by azriel91** [comment]
now we'll have a less embarrassing coverage % :stuck_out_tongue:

**Comment by yl54** [comment]
small steps
